edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/EdgeHubCertificates.cs

```diff
@@ -33,64 +36,92 @@ public class EdgeHubCertificates
+                string edgeHubDockerCaChainCertPath = configuration.GetValue<string>(Constants.ConfigKey.EdgeHubServerCAChainCertificateFile);
+                string edgeHubConnectionString = configuration.GetValue<string>(Constants.ConfigKey.IotHubConnectionString);
 
-                certificates = CertificateHelper.GetServerCertificateAndChainFromFile(edgeHubDevCertPath, edgeHubDevPrivateKeyPath);
-                IEnumerable<X509Certificate2> trustBundle = CertificateHelper.ParseTrustedBundleFromFile(edgeHubDevTrustBundlePath);
+                if (string.IsNullOrEmpty(edgeHubConnectionString))
+                {
+                    // When connection string is not set it is edged mode as iotedgd is expected to set this.
+                    // In this case we reach out to the iotedged via the workload interface.
+                    (X509Certificate2 ServerCertificate, IEnumerable<X509Certificate2> CertificateChain) certificates;
 
-                result = new EdgeHubCertificates(
-                    certificates.ServerCertificate,
-                    certificates.CertificateChain?.ToList(),
-                    trustBundle?.ToList(),
-                    Option.None<X509Certificate2>());
-            }
-            else if (!string.IsNullOrEmpty(edgeHubDockerCertPFXPath) &&
-                     !string.IsNullOrEmpty(edgeHubDockerCaChainCertPath))
-            {
-                // If no connection string was set and we use iotedge devdiv style certificates for development
-                List<X509Certificate2> certificateChain = CertificateHelper.GetServerCACertificatesFromFile(edgeHubDockerCaChainCertPath)?.ToList();
-                result = new EdgeHubCertificates(new X509Certificate2(edgeHubDockerCertPFXPath), certificateChain, new List<X509Certificate2>(), Option.None<X509Certificate2>());
-            }
-            else
-            {
-                throw new InvalidOperationException("Edge Hub certificate files incorrectly configured");
+                    var workloadUri = new Uri(configuration.GetValue<string>(Constants.ConfigKey.WorkloadUri));
+                    string edgeHubHostname = configuration.GetValue<string>(Constants.ConfigKey.EdgeDeviceHostName);
+                    string moduleId = configuration.GetValue<string>(Constants.ConfigKey.ModuleId);
+                    string generationId = configuration.GetValue<string>(Constants.ConfigKey.ModuleGenerationId);
+                    string edgeletApiVersion = configuration.GetValue<string>(Constants.ConfigKey.WorkloadAPiVersion);
+                    DateTime expiration = DateTime.UtcNow.AddDays(Constants.CertificateValidityDays);
+
+                    certificates = await CertificateHelper.GetServerCertificatesFromEdgelet(workloadUri, edgeletApiVersion, Constants.WorkloadApiVersion, moduleId, generationId, edgeHubHostname, expiration);
```

Comment:
Can we instead move this retry logic to CertificateHelper.GetServerCertificatesFromEdgelet? If the server cert it gets has no private key, then we try to get the cert again.

Reply:
the 'GetServerCertificateAndChainFromFile' call is also affected. I can move the logic into 'ParseCertificateAndKey()' I think